src/inference/mh.jl

```diff
@@ -132,78 +150,41 @@ Returns two `NamedTuples`. The first `NamedTuple` has symbols as keys and distri
 The second `NamedTuple` has model symbols as keys and their stored values as values.
 """
 function dist_val_tuple(spl::Sampler{<:MH})
-    vns = _getvns(spl.state.vi, spl)
-    dt = _dist_tuple(spl.state.vi.metadata, spl.alg.proposals, spl.state.vi, vns)
-    vt = _val_tuple(spl.state.vi.metadata, spl.state.vi, vns)
+    vi = spl.state.vi
+    vns = _getvns(vi, spl)
+    dt = _dist_tuple(spl.alg.proposals, vi, vns)
+    vt = _val_tuple(vi, vns)
     return dt, vt
 end
 
-@generated function _val_tuple(metadata::NamedTuple, vi::VarInfo, vns::NamedTuple{names}) where {names}
-    length(names) === 0 && return :(NamedTuple())
-    expr = Expr(:tuple)
-    map(names) do f
-        push!(expr.args, Expr(:(=), f, :(scalar_map(vi, metadata.$f.vns))))
-    end
-    return expr
+function _val_tuple(
```

Comment:
Any reason the generated functions were removed? My understanding was that they were significantly faster for named tuple generation (perhaps a faulty understanding).

Reply:
I haven't benchmarked it, I guess I should do that :stuck_out_tongue: I just usually try to avoid generated functions (for reasons such as mentioned in [this discourse discussion](https://discourse.julialang.org/t/why-not-use-generated/29531/4?u=devmotion)), and it didn't feel necessary to use generated functions here since the same functionality could be achieved using `map`.

Reply:
I know @mohamed82008 has worked a lot with these. Mohamed, do you know how well generated functions benchmark when returning NamedTuples?

Reply:
I noticed that it's not possible to compare the old and the new version in a fair way since the logic changed (and is type unstable now since we want to unpack single distributions that are wrapped in an array but must not unpack arrays with multiple distributions). Instead, I compared a simple (but somewhat similar) example
```julia
@generated function square_generated(a::NamedTuple{names}, b::NamedTuple) where names
    length(names) === 0 && return :(NamedTuple())
    expr = Expr(:tuple)
    map(names) do f
        push!(expr.args, Expr(:(=), f, :(b.$f^2)))
    end
    return expr
end

function square_map(a::NamedTuple{names}, b::NamedTuple) where names
    return NamedTuple{names}(
        map(names) do f
            b[f]^2
        end
    )
end

function square_broadcast(a::NamedTuple{names}, b::NamedTuple) where names
    return NamedTuple{names}(
        getfield.(Ref(b), names).^2
    )
end
```
with
```julia
using BenchmarkTools
using Random

names = [Symbol.(randstring(4)) for _ in 1:100]
a = NamedTuple{Tuple(names)}(rand(100))
b = NamedTuple{Tuple(names[randperm(100)])}(rand(100)
```
I got
```julia
julia> @btime square_generated($a, $b);
  62.283 ns (0 allocations: 0 bytes)

julia> @btime square_map($a, $b);
  17.847 μs (303 allocations: 85.28 KiB)

julia> @btime square_broadcast($a, $b);
  14.124 μs (200 allocations: 81.25 KiB)
```
so at least in this example clearly the implementation with a generated function is the most efficient one, by far.

I'll revert it to an implementation with generated functions! :+1:

Reply:
I'm honestly surprised at how crazy good the generated function is -- I expected maybe twice as fast as the mapped version, but this is just crazy. Plus, no allocations!